src/update_worker.rs

```diff
@@ -0,0 +1,141 @@
+use anyhow::ensure;
+use reqwest::{header, Client, Url};
+use std::{
+    fmt::{Display, Formatter},
+    path::Path,
+};
+
+#[derive(Debug)]
+pub struct DiffError {
+    pub worker_id: usize,
+    pub update_id: usize,
+    pub update: String,
+    pub expected: String,
+    pub actual: String,
+}
+
+fn format_insert_or_delete_data(f: &mut Formatter<'_>, q: &str) -> std::fmt::Result {
+    let (head, body) = q.split_once("{ <").unwrap();
+    let (body, _) = body.rsplit_once(". }").unwrap();
+
+    writeln!(f, "{head}{{")?;
+
+    for triple in body.split(". <") {
+        writeln!(f, "    <{triple}.")?;
+    }
+
+    writeln!(f, "}}")?;
+    Ok(())
+}
+
+impl Display for DiffError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Unexpected result from worker {} at update {}\n\nquery:\n",
+            self.worker_id, self.update_id
+        )?;
+
+        if self.update.starts_with("INSERT DATA") || self.update.starts_with("DELETE DATA") {
+            format_insert_or_delete_data(f, &self.update)?;
+        } else {
+            writeln!(f, "{}", self.update)?;
+        }
+
+        writeln!(f, "\nexpected:\n{}", self.expected)?;
+        writeln!(f, "\nactual:\n{}", self.actual)?;
+
+        writeln!(f, "\ndiff:\n{}", prettydiff::diff_lines(&self.expected, &self.actual))
+    }
+}
+
+pub struct UpdateWorker {
+    id: usize,
+    query_endpoint: Url,
+    update_endpoint: Url,
+    client: Client,
+    queries: Vec<(Subject, Query, DbState)>,
+}
+
+impl UpdateWorker {
+    fn make_verify_query(subject: &str) -> String {
+        format!("CONSTRUCT {{ {subject} ?p ?o }} WHERE {{ {subject} ?p ?o . FILTER(!isBlank(?o)) }}")
+    }
+
+    pub fn new(id: usize, base_dir: &Path, query_endpoint: Url, update_endpoint: Url) -> anyhow::Result<Self> {
+        let mut queries = Vec::new();
+
+        for op in 1.. {
+            let subject = base_dir.join(format!("op_{op}.txt"));
+            let update = base_dir.join(format!("op_{op}.ru"));
+            let update_result = base_dir.join(format!("op_{op}.nt"));
+
+            if !subject.exists() && !update.exists() && !update_result.exists() {
+                break;
+            }
+
+            let subject = std::fs::read_to_string(subject)?;
+            let update = std::fs::read_to_string(update)?;
+            let mut update_result: Vec<_> = std::fs::read_to_string(update_result)?
+                .lines()
+                .map(ToOwned::to_owned)
+                .collect();
+
+            update_result.sort();
+
+            queries.push((subject, update, update_result));
+        }
+
+        Ok(Self { id, query_endpoint, update_endpoint, client: Client::new(), queries })
+    }
+
+    async fn read_current_state(&self, subject: &str) -> anyhow::Result<DbState> {
+        let state = self
+            .client
+            .get(self.query_endpoint.clone())
+            .query(&[("query", Self::make_verify_query(subject))])
```

Comment:
th Accept header should be set to "application/n-triples" to ensure that the endpoint delivers the expected format.